debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotMySQLUtility.java

```diff
@@ -0,0 +1,543 @@
+                            // ...
+                            final Table table = schema.tableFor(tableId);
+                            final int numColumns = table.columns().size();
+                            final Object[] row = new Object[numColumns];
+                            while (rs.next()) {
+                                for (int i = 0, j = 1; i != numColumns; ++i, ++j) {
+                                    row[i] = rs.getObject(j);
+                                }
+                                recorder.recordRow(recordMaker, row, bufferingStartTs); // has no row number!
+                                ++rowNum;
+                                if (rowNum % 10_000 == 0) {
+                                    long stop = context.clock().currentTimeInMillis();
+                                    logger.info("{} of {} rows scanned from table '{}' after {}",
+                                                rowNum, rowCountStr, tableId, Strings.duration(stop - start));
+                                }
+                            }
+
+                            totalRowCount.addAndGet(rowNum);
+                        } catch (InterruptedException e) {
+                            Thread.interrupted();
+                            // We were not able to finish all rows in all tables ...
+                            logger.info("Stopping the snapshot due to thread interruption");
+                            interrupted.set(true);
+                        }
+                    });
+                } finally {
+                    metrics.completeTable();
+                    if (interrupted.get()) break;
+                }
+            }
+            ++completedCounter;
+        }
+
+        // We've copied all of the tables and we've not yet been stopped, but our buffer holds onto the
+        // very last record. First mark the snapshot as complete and then apply the updated offset to
+        // the buffered record ...
+        context.source().markLastSnapshot();
+        long stop = context.clock().currentTimeInMillis();
+        try {
+            bufferedRecordQueue.flush(this::replaceOffset);
+            logger.info("scanned {} rows in {} tables in {}",
+                        totalRowCount, tableIdsToBuffer.size(), Strings.duration(stop - startScan));
+        } catch (InterruptedException e) {
+            Thread.interrupted();
+            // We were not able to finish all rows in all tables ...
+            logger.info("aborting the snapshot after {} rows in {} of {} tables {}",
+                        totalRowCount, completedCounter, tableIdsToBuffer.size(), Strings.duration(stop - startScan));
+            interrupted.set(true);
+        }
+    }
+
+    /**
+     * Complete the current transaction, and unlock any tables that are still locked.
+     */
+    public void completeTransaction() throws SQLException {
```

Comment:
Please consider renaming - you are completing TX AND releasing the table locks

Reply:
renamed